crypto/cmp/cmp_util.c

```diff
@@ -0,0 +1,404 @@
+        return cnt;
+    return 0;
+}
+
+/*
+ * auxiliary function for incrementally reporting texts via the error queue
+ */
+
+void OSSL_CMP_add_error_txt(const char *separator, const char *txt)
+{
+    const char *file;
+    int line;
+    const char *data;
+    int flags;
+    unsigned long err = ERR_peek_last_error();
+
+    if (separator == NULL)
+        separator = "";
+    if (err == 0)
+        ERR_PUT_error(ERR_LIB_CMP, 0, err, "", 0);
+
+#define MAX_DATA_LEN 4096-100 /* workaround for ERR_print_errors_cb() limit */
+    do {
+        const char *curr, *next;
+        int len;
+        char *tmp;
+
+        ERR_peek_last_error_line_data(&file, &line, &data, &flags);
+        if ((flags & ERR_TXT_STRING) == 0) {
+            data = "";
+            separator = "";
+        }
+        len = (int)strlen(data);
+        curr = next = txt;
+        while (*next != '\0'
+                   && len + strlen(separator) + (next - txt) < MAX_DATA_LEN) {
+            curr = next;
+            if (*separator != '\0') {
+                next = strstr(curr, separator);
+                if (next != NULL)
+                    next += strlen(separator);
+                else
+                    next = curr + strlen(curr);
+            } else
+                next = curr + 1;
+        }
+        if (*next != '\0') { /* split error msg if error data gets too long */
+            if (curr != txt) {
+                tmp = OPENSSL_strndup(txt, curr - txt);
+                ERR_add_error_data(2, separator, tmp);
+                OPENSSL_free(tmp);
+            }
+            ERR_PUT_error(ERR_LIB_CMP, 0 /* func */, err, file, line);
+            txt = curr;
+        } else {
+            ERR_add_error_data(2, separator, txt);
+            txt = next;
+        }
+    } while (*txt != '\0');
+}
+
+/* this is similar to ERR_print_errors_cb, but uses the CMP-specific cb type */
+void OSSL_CMP_print_errors_cb(OSSL_cmp_log_cb_t log_fn)
+{
+    unsigned long err;
+    char component[256];
+    char msg[4096];
+    const char *file, *data;
+    int line, flags;
+
+    if (log_fn == NULL) {
+#ifndef OPENSSL_NO_STDIO
+        BIO *bio_err = BIO_new_fp(stderr, BIO_NOCLOSE);
+
+        ERR_print_errors(bio_err);
+        BIO_free(bio_err);
+#endif
+        return;
+    }
+
+    while ((err = ERR_get_error_line_data(&file, &line, &data, &flags)) != 0) {
+        if (!(flags & ERR_TXT_STRING))
+            data = NULL;
+        BIO_snprintf(component, sizeof(component), "OpenSSL:%s",
+                     /* ERR_lib_error_string(err), */
+                     ERR_func_error_string(err));
+        BIO_snprintf(msg, sizeof(msg), "%s%s%s", ERR_reason_error_string(err),
+                     data == NULL ? "" : " : ", data == NULL ? "" : data);
+        if (log_fn(component, file, line, OSSL_CMP_LOG_ERR, msg) <= 0)
+            break;              /* abort outputting the error report */
+    }
+}
+
+/*
+ * functions manipulating lists of certificates etc.
```

Comment:
You are right, these functions actually are independent of CMP, and yet we need them here and there and we must export them just because we'll need them in `apps/cmp.c`, which going to be introduced in chunk 10.
For these reasons I decided to collect them in this utility code file.

Do you think that some or all of them would better be moved to some other (core) parts of the crypto lib?
If so, I suggest you let us know in which source file(s) to put them with which names,
and we can easily do this as part of this PR.

Reply:
@mattcaswell, did you see this open question? 
Or is this ok at least for the time being?

Reply:
I am reluctant to add lots of new public API helper functions without a strong reason. Our API is big enough as it is and I prefer to try and keep it down to the bare essentials if at all possible. That might mean we have some duplicated code in the apps. I'm ok with leaving these functions here if they are not public (in which case they will need renaming to make the names lowercase).

Reply:
All right, I've minimized the util functions and made them internal.
Doing so I had to duplicate code into apps/cmp.c (which is to be addded in a later chunk).